pkg/modifycontroller/controller_test.go

```diff
@@ -154,77 +110,152 @@ func TestModifyPVC(t *testing.T) {
+			name:          "Should NOT modify if PVC's PV not created yet",
+			pvc:           pvcWithUncreatedPV,
+			pv:            basePV,
+			callCSIModify: false,
+		},
+	}
 
-			stopCh := make(chan struct{})
-			informerFactory.Start(stopCh)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			client := csi.NewMockClient(testDriverName, true, true, true, true, true, false)
 
-			ctx := context.TODO()
+			initialObjects := []runtime.Object{test.pvc, test.pv, testVacObject, targetVacObject}
+			ctrlInstance, ctx := setupFakeK8sEnvironment(t, client, initialObjects)
 			defer ctx.Done()
-			go controller.Run(1, ctx)
-
-			for _, obj := range initialObjects {
-				switch obj.(type) {
-				case *v1.PersistentVolume:
-					pvInformer.Informer().GetStore().Add(obj)
-				case *v1.PersistentVolumeClaim:
-					pvcInformer.Informer().GetStore().Add(obj)
-				case *storagev1beta1.VolumeAttributesClass:
-					vacInformer.Informer().GetStore().Add(obj)
-				default:
-					t.Fatalf("Test %s: Unknown initalObject type: %+v", test.name, obj)
-				}
-			}
-
-			time.Sleep(time.Second * 2)
 
-			_, _, err, _ = ctrlInstance.modify(test.pvc, test.pv)
+			err := ctrlInstance.syncPVC(pvcNamespace + "/" + pvcName)
+			if err != nil {
+				t.Errorf("for %s, unexpected error: %v", test.name, err)
+			}
 
-			if test.expectFailure && err == nil {
-				t.Errorf("for %s expected error got nothing", test.name)
+			modifyCallCount := client.GetModifyCount()
+			if test.callCSIModify && modifyCallCount == 0 {
+				t.Fatalf("for %s: expected csi modify call, no csi modify call was made", test.name)
 			}
 
-			if !test.expectFailure {
-				if err != nil {
-					t.Errorf("for %s, unexpected error: %v", test.name, err)
-				}
+			if !test.callCSIModify && modifyCallCount > 0 {
+				t.Fatalf("for %s: expected no csi modify call, received csi modify request", test.name)
 			}
 		})
 	}
 }
+
+// setupFakeK8sEnvironment creates fake K8s environment and starts Informers and ModifyController
+func setupFakeK8sEnvironment(t *testing.T, client *csi.MockClient, initialObjects []runtime.Object) (*modifyController, context.Context) {
+	t.Helper()
+
+	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.VolumeAttributesClass, true)
+
+	/* Create fake kubeClient, Informers, and ModifyController */
+	kubeClient, informerFactory := fakeK8s(initialObjects)
+	pvInformer := informerFactory.Core().V1().PersistentVolumes()
+	pvcInformer := informerFactory.Core().V1().PersistentVolumeClaims()
+	vacInformer := informerFactory.Storage().V1beta1().VolumeAttributesClasses()
+
+	driverName, _ := client.GetDriverName(context.TODO())
+
+	csiModifier, err := modifier.NewModifierFromClient(client, 15*time.Second, kubeClient, informerFactory, false, driverName)
+	if err != nil {
+		t.Fatalf("Test %s: Unable to create modifier: %v", t.Name(), err)
+	}
+
+	controller := NewModifyController(driverName,
+		csiModifier, kubeClient,
+		0, false, informerFactory,
```

Comment:
Note that I have set resync period to 0 because shared_informer does not support resyncing:

`W1023 19:04:49.983836   13491 shared_informer.go:569] The specified resyncPeriod 1s is invalid because this shared informer doesn't support resyncing`

This also let us cut out any sleeps in unit tests (instant tests!)